File: Data Structures & Algorithms/binary-tree-diameter/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        Ans fa = helper(root);
        return max(fa.md-1, fa.sa);
    }

private:
    class Ans {
        public: 
            int md;
            int sa;
            Ans(int x, int y): md(x), sa(y) {
                
            };
    };

    Ans helper(TreeNode* root) {
        if (root == nullptr) {
            return Ans(0,0);
        }
        Ans la = helper(root->left);
        Ans ra = helper(root->right);
        return Ans(max(la.md, ra.md) + 1, max(max(la.sa, ra.sa), la.md+ra.md));
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int pv;
        int qv;
        if (p.val < q.val) {
            pv = p.val;
            qv = q.val;
        } else {
            pv = q.val;
            qv = p.val;
        }
        while (root != null) {
            boolean a = (root.val == pv || root.val > pv);
            boolean b = (root.val == qv || root.val < qv);
            if (a && b) {
                return root;
            }
            if (a) {
                root = root.left;
            } else {
                root = root.right;
            }
        }
        return null;
    }
}
